[NEMO-328] Refactor IRDAG

johnyangk opened the pull request:
JIRA: [NEMO-328: Refactor IRDAG](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-328)

**Major changes:**
- Modifies the Pass interface to use IRDAG, rather than directly using DAG
- Modifies reshaping passes (e.g., LargeShuffle, Skew) to use the IRDAG reshaping methods

**Minor changes to note:**
- Merges multiple annotating LargeShufflePasses into a single LargeShuffleAnnotatingPass

**Tests for the changes:**
- Existing tests pass

**Other comments:**
- N/A

Closes #190

sanha commented on compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/SkewReshapingPass.java:
```diff
@@ -47,17 +41,13 @@
- * 1) Task-level statistic collection is done via vertex with {@link MetricCollectTransform}
- * 2) Stage-level statistic aggregation is done via vertex with {@link AggregateMetricTransform}
- * inserted before shuffle edges.
+ * We insert a {@link MessageBarrierVertex for each shuffle edge}
```
We **must** optimize multiple shuffle edges at once if a vertex receives many shuffle edges. Please check [327](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-327?filter=allopenissues).

wonook commented on compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/composite/LargeShuffleCompositePass.java:
```diff
@@ -33,16 +33,8 @@
-        new LargeShuffleDataPersistencePass(),
-        new LargeShuffleResourceSlotPass()
+        new LargeShuffleReshapingPass(),
+        new LargeShuffleAnnotatingPass()
```
Should other composite passes refactored this way as well?

wonook commented on common/src/main/java/org/apache/nemo/common/ir/IRDAG.java:
```diff
@@ -0,0 +1,286 @@
+   * Get the current underlying DAG for direct access.
+   * @return underlying DAG.
+   */
+  public DAG<IRVertex, IREdge> getCurrentDAGSnapshot() {
```
How is this a 'snapshot'?

wonook commented on common/src/main/java/org/apache/nemo/common/ir/IRDAG.java:
```diff
@@ -0,0 +1,286 @@
+   * "Unsafe" direct reshaping (semantic-preserving is not guaranteed).
+   * @param unsafeReshaping a function that directly reshapes the underlying DAG.
+   */
+  public void unSafeDirectReshaping(final Function<DAG<IRVertex, IREdge>, DAG<IRVertex, IREdge>> unsafeReshaping) {
```
Maybe something like 'reshapeUnsafely' or 'applyUnsafeDirectReshaping' or something like that? Unsafe is a single word by the way. I think it would be better to keep the names of method that perform a specific action as verbs. Could you double check on this throughout your PR?

wonook replied:
DAG<IRVertex, IREdge> seems to be still used here. I don't see a reason for using both forms.

johnyangk commented:
Thanks for the comments. I've updated the code.

@sanha 
I've modified IRDAG#insert to receive multiple IREdges so the #189 continues to function as intended. I've run NetworkTraceAnalysisITCase#testDataSkew and verified the DAG topology.

@wonook 
I've addressed your comments except for IRDAG#unSafeDirectReshaping, because the loop-related passes require direct access to the underlying DAG. It'd be good to handle this issue with a separate PR ([JIRA](https://issues.apache.org/jira/projects/NEMO/issues/NEMO-330)). There also remains around 15 instances of DAG<IRVertex, IREdge> outside of IRDAG, mostly in loop-related passes and 2 in Spark-related classes for building new IRDAGs from RDD applications. Other than those instances, I've changed the type to IRDAG for consistency.

wonook commented:
@sanha Anything else you want to check?

johnyangk commented:
@wonook Thanks. I've renamed as per your suggestions.

johnyangk commented on compiler/optimizer/src/main/java/org/apache/nemo/compiler/optimizer/pass/compiletime/reshaping/SkewReshapingPass.java:
```diff
@@ -66,186 +59,66 @@ public SkewReshapingPass() {
-            final IREdge edgeToOriginalDstV =
-              new IREdge(edge.getPropertyValue(CommunicationPatternProperty.class).get(), edge.getSrc(), v);
-            edge.copyExecutionPropertiesTo(edgeToOriginalDstV);
-            edgeToOriginalDstV.setPropertyPermanently(MetricCollectionProperty.of(metricCollectionId));
```
@sanha This has been moved to IRDAG#insert

https://github.com/apache/incubator-nemo/pull/190/files#diff-969693370fcffe6720d48d175502c466R210
